ccl/migrationccl/migrationsccl: TestRecordsBasedRegistryMigration failed

ccl/migrationccl/migrationsccl.TestRecordsBasedRegistryMigration [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4010770&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4010770&tab=artifacts#/) on master @ [5ac733bb4927020bc1c52da24b2591742fde8e1f](https://github.com/cockroachdb/cockroach/commits/5ac733bb4927020bc1c52da24b2591742fde8e1f):

Fatal error:

```
panic: use of Span after Finish. Span: dist sender. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

```
Stack: 

```
goroutine 140789 [running]:
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0xc0001fecb0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:99 +0x134
github.com/cockroachdb/cockroach/pkg/util/tracing.WithParent(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span_options.go:180
github.com/cockroachdb/cockroach/pkg/util/tracing.EnsureForkSpan({0x4c30978, 0xc0059eaf00}, 0xc0001fecb0, {0xc0029f0150, 0x29})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/tracer.go:1226 +0x105
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleClosedTimestampUpdateRaftMuLocked(0xc000f75500, {0x4c30978, 0xc0059eaf00}, {0xc005b81320, 0x35885a0, 0xc0})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_rangefeed.go:663 +0x325
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).registerWithRangefeedRaftMuLocked(0xc000f75500, {0x4c30978, 0xc0059eaf00}, {{0xc0013ab7a0, 0x1, 0x8}, {0xc0013ab7a8, 0x1, 0x8}}, {0x16c5828034ed5068, ...}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_rangefeed.go:427 +0x731
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).rangeFeedWithRangeID(0xc000f75500, 0x10, 0xc004d6f0e0, {0x4c99ba0, 0xc002aca920})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_rangefeed.go:238 +0x635
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).RangeFeed(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_rangefeed.go:146
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).RangeFeed(0xc000cea000, 0xc004d6f0e0, {0x4c99ba0, 0xc002aca920})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store.go:2986 +0xdc
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).RangeFeed(0x0, 0xc004d6f0e0, {0x4c99ba0, 0xc002aca920})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores.go:216 +0xe8
github.com/cockroachdb/cockroach/pkg/server.(*Node).RangeFeed(0x4c99ba0, 0xf0f000000000000, {0x4c99ba0, 0xc002aca920})
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:1157 +0x38
github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.RangeFeed.func1()
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:671 +0xbb
created by github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.RangeFeed
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:669 +0x18f
```
<details><summary>Log preceding fatal error</summary>
<p>

```
I211230 10:45:26.371769 8638 3@vendor/github.com/cockroachdb/pebble/event.go:636  [n?,pebble,s?] 151  [JOB 1] MANIFEST created 000001
I211230 10:45:26.371907 8638 3@vendor/github.com/cockroachdb/pebble/event.go:664  [n?,pebble,s?] 152  [JOB 1] WAL created 000002
I211230 10:45:26.372154 115094 3@vendor/github.com/cockroachdb/pebble/event.go:656  [n?,pebble,stemp] 153  [JOB 2] all initial table stats loaded
I211230 10:45:26.405108 8638 1@server/server.go:1053  [n?] 154  monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I211230 10:45:26.407186 8638 1@server/server.go:1096  [n1] 155  Sleeping till wall time 1640861126407172065 to catches up to 1640861126905090477 to ensure monotonicity. Delta: 497.918412ms
I211230 10:45:26.407639 116086 1@server/server.go:1620  [n1] 156  connecting to gossip network to verify cluster ID "f22a4b12-7124-4539-8392-fa7c529d125a"
W211230 10:45:26.905683 8638 1@gossip/gossip.go:1472  [n1] 157  no addresses found; use --join to specify a connected node
I211230 10:45:26.905866 8638 gossip/gossip.go:401  [n1] 158  NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:34359" > attrs:<> locality:<tiers:<key:"region" value:"test" > tiers:<key:"dc" value:"dc1" > > ServerVersion:<major_val:21 minor_val:2 patch:0 internal:34 > build_tag:"v22.1.0-alpha.00000000-2469-g5ac733bb49" started_at:1640861126905837909 cluster_name:"" sql_address:<network_field:"tcp" address_field:"127.0.0.1:38685" >
I211230 10:45:26.918260 8638 server/node.go:451  [n1] 160  initialized store s1
I211230 10:45:26.918203 116086 1@server/server.go:1623  [n1] 159  node connected via gossip
I211230 10:45:26.918371 8638 kv/kvserver/stores.go:250  [n1] 161  read 0 node addresses from persistent storage
I211230 10:45:26.918873 8638 server/node.go:528  [n1] 162  started with engine type 2
I211230 10:45:26.919024 8638 server/node.go:530  [n1] 163  started with attributes []
I211230 10:45:26.919998 8638 1@server/server.go:1751  [n1] 164  starting https server at 127.0.0.1:36563 (use: 127.0.0.1:36563)
I211230 10:45:26.920157 8638 1@server/server.go:1756  [n1] 165  starting postgres server at 127.0.0.1:38685 (use: 127.0.0.1:38685)
I211230 10:45:26.920225 8638 1@server/server.go:1758  [n1] 166  starting grpc server at 127.0.0.1:34359
I211230 10:45:26.920273 8638 1@server/server.go:1759  [n1] 167  advertising CockroachDB node at 127.0.0.1:34359
I211230 10:45:26.934051 8638 1@util/log/event_log.go:32  [n1] 168 ={"Timestamp":1640861126934047346,"EventType":"node_restart","NodeID":1,"StartedAt":1640861126905837909,"LastUp":1640861126203631368}
I211230 10:45:26.934260 8638 sql/sqlliveness/slinstance/slinstance.go:300  [n1] 169  starting SQL liveness instance
I211230 10:45:26.943125 8638 sql/sqlstats/persistedsqlstats/scheduled_job_monitor.go:145  [n1] 170  cannot create sql stats scheduled compaction job because current cluster version is too low
I211230 10:45:26.944048 120973 sql/temporary_schema.go:515  [n1] 171  running temporary object cleanup background job
I211230 10:45:26.944224 119754 sql/sqlliveness/slstorage/slstorage.go:406  [n1] 172  inserted sqlliveness session e19846e5c7844908acee663f66979248
I211230 10:45:26.944335 121019 sql/sqlstats/persistedsqlstats/provider.go:137  [n1] 173  starting sql-stats-worker with initial delay: 1h5m10.562881865s
I211230 10:45:26.944659 119754 sql/sqlliveness/slinstance/slinstance.go:187  [n1] 174  created new SQL liveness session e19846e5c7844908acee663f66979248
I211230 10:45:26.945904 8638 server/server_sql.go:1161  [n1] 175  done ensuring all necessary startup migrations have run
I211230 10:45:26.946825 121507 jobs/job_scheduler.go:357  [n1] 176  waiting 4m0s before scheduled jobs daemon start
I211230 10:45:26.948063 121658 server/auto_upgrade.go:44  [n1] 177  auto upgrade disabled by testing
I211230 10:45:26.970166 120973 sql/temporary_schema.go:559  [n1] 178  found 0 temporary schemas
I211230 10:45:26.970236 120973 sql/temporary_schema.go:562  [n1] 179  early exiting temporary schema cleaner as no temporary schemas were found
I211230 10:45:26.970271 120973 sql/temporary_schema.go:563  [n1] 180  completed temporary object cleanup job
I211230 10:45:26.970303 120973 sql/temporary_schema.go:646  [n1] 181  temporary object cleaner next scheduled to run at 2021-12-30 11:15:26.943994934 +0000 UTC
I211230 10:45:26.981686 8638 1@server/server.go:2443  [n1] 182  serving sql connections
I211230 10:45:26.983530 8638 testutils/testcluster/testcluster.go:1158  [-] 183  WaitForFullReplication
I211230 10:45:26.983581 8638 testutils/testcluster/testcluster.go:1162  [-] 184  WaitForFullReplication took: 253ns
I211230 10:45:26.983681 8638 2@rpc/context.go:1107  [n1] 185  dialing n1: 127.0.0.1:34359 (default)
I211230 10:45:27.006558 130364 kv/kvserver/replica_rangefeed.go:652  [n1,rangefeed=sql_instances,s1,r42/1:/Table/4{6-7}] 186  RangeFeed closed timestamp is empty
I211230 10:45:27.027198 127487 sql/catalog/lease/lease.go:1284  [n1] 187  released orphaned lease: {id:31 version:1 expiration:{Time:{wall:140338000 ext:63776458232 loc:<nil>}}}
I211230 10:45:27.042086 127398 sql/catalog/lease/lease.go:1284  [n1] 188  released orphaned lease: {id:6 version:1 expiration:{Time:{wall:549126000 ext:63776458217 loc:<nil>}}}
I211230 10:45:27.045467 127429 sql/catalog/lease/lease.go:1284  [n1] 189  released orphaned lease: {id:20 version:1 expiration:{Time:{wall:689979000 ext:63776458228 loc:<nil>}}}
I211230 10:45:27.046675 127403 sql/catalog/lease/lease.go:1284  [n1] 190  released orphaned lease: {id:12 version:1 expiration:{Time:{wall:460715000 ext:63776458230 loc:<nil>}}}
I211230 10:45:27.049906 133647 sql/catalog/lease/lease.go:1284  [n1] 191  released orphaned lease: {id:35 version:1 expiration:{Time:{wall:348809000 ext:63776458230 loc:<nil>}}}
I211230 10:45:27.051425 127408 sql/catalog/lease/lease.go:1284  [n1] 192  released orphaned lease: {id:15 version:1 expiration:{Time:{wall:206683000 ext:63776458235 loc:<nil>}}}
I211230 10:45:27.056484 134656 sql/catalog/lease/lease.go:1284  [n1] 193  released orphaned lease: {id:40 version:1 expiration:{Time:{wall:964364000 ext:63776458233 loc:<nil>}}}
I211230 10:45:27.058430 135344 sql/catalog/lease/lease.go:1284  [n1] 194  released orphaned lease: {id:52 version:1 expiration:{Time:{wall:38929000 ext:63776458235 loc:<nil>}}}
I211230 10:45:27.058479 135162 sql/catalog/lease/lease.go:1284  [n1] 195  released orphaned lease: {id:50 version:1 expiration:{Time:{wall:711457000 ext:63776458229 loc:<nil>}}}
I211230 10:45:27.103681 8638 testutils/testcluster/testcluster.go:106  [-] 196  TestCluster quiescing nodes
W211230 10:45:27.103925 121066 jobs/registry.go:859  [n1] 197  canceling all adopted jobs due to stopper quiescing
I211230 10:45:27.104061 126807 2@rpc/context.go:742  [n1] 198  closing [{n1: 127.0.0.1:34359 (default)} {n0: 127.0.0.1:34359 (default)}]
W211230 10:45:27.104124 119754 sql/sqlliveness/slinstance/slinstance.go:224  [n1] 199  exiting heartbeat loop
I211230 10:45:27.104431 114041 2@rpc/context.go:742  [n1] 200  closing [{n1: 127.0.0.1:34359 (default)}]
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRecordsBasedRegistryMigration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/rpc/context.go
+++ b/pkg/rpc/context.go
@@ -660,6 +660,7 @@ func (a internalClientAdapter) RangeFeed(
 	ctx context.Context, args *roachpb.RangeFeedRequest, _ ...grpc.CallOption,
 ) (roachpb.Internal_RangeFeedClient, error) {
 	ctx, cancel := context.WithCancel(ctx)
+	ctx, sp := tracing.ChildSpan(ctx, "/cockroach.roachpb.Internal/RangeFeed")
 	rfAdapter := rangeFeedClientAdapter{
 		respStreamClientAdapter: makeRespStreamClientAdapter(ctx),
 	}
@@ -668,6 +669,7 @@ func (a internalClientAdapter) RangeFeed(
 	args.AdmissionHeader.SourceLocation = roachpb.AdmissionHeader_LOCAL
 	go func() {
 		defer cancel()
+		defer sp.Finish()
 		err := a.server.RangeFeed(args, rfAdapter)
 		if err == nil {
 			err = io.EOF
@@ -704,12 +706,14 @@ func (a internalClientAdapter) GossipSubscription(
 	ctx context.Context, args *roachpb.GossipSubscriptionRequest, _ ...grpc.CallOption,
 ) (roachpb.Internal_GossipSubscriptionClient, error) {
 	ctx, cancel := context.WithCancel(ctx)
+	ctx, sp := tracing.ChildSpan(ctx, "/cockroach.roachpb.Internal/GossipSubscription")
 	gsAdapter := gossipSubscriptionClientAdapter{
 		respStreamClientAdapter: makeRespStreamClientAdapter(ctx),
 	}
 
 	go func() {
 		defer cancel()
+		defer sp.Finish()
 		err := a.server.GossipSubscription(args, gsAdapter)
 		if err == nil {
 			err = io.EOF

